state/txindex/kv/kv.go

```diff
@@ -46,13 +59,321 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
 	return txResult, nil
 }
 
-// AddBatch writes a batch of transactions into the TxIndex storage.
+// AddBatch indexes a batch of transactions using the given list of tags.
 func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
 	storeBatch := txi.store.NewBatch()
+
 	for _, result := range b.Ops {
-		rawBytes := wire.BinaryBytes(&result)
-		storeBatch.Set(result.Tx.Hash(), rawBytes)
+		hash := result.Tx.Hash()
+
+		// index tx by tags
+		for _, tag := range result.Result.Tags {
+			if stringInSlice(tag.Key, txi.tagsToIndex) {
+				storeBatch.Set(keyForTag(tag, result), hash)
+			}
+		}
+
+		// index tx by hash
+		rawBytes := wire.BinaryBytes(result)
+		storeBatch.Set(hash, rawBytes)
 	}
+
 	storeBatch.Write()
 	return nil
 }
+
+// Index indexes a single transaction using the given list of tags.
+func (txi *TxIndex) Index(result *types.TxResult) error {
+	batch := txindex.NewBatch(1)
+	err := batch.Add(result)
+	if err != nil {
+		return errors.Wrap(err, "failed to add tx result to batch")
+	}
+	return txi.AddBatch(batch)
+}
+
+// Search performs a search using the given query. It breaks the query into
+// conditions (like "tx.height > 5"). For each condition, it queries the DB
+// index. One special use cases here: (1) if "tx.hash" is found, it returns tx
+// result for it (2) for range queries it is better for the client to provide
+// both lower and upper bounds, so we are not performing a full scan. Results
+// from querying indexes are then intersected and returned to the caller.
+func (txi *TxIndex) Search(q *query.Query) ([]*types.TxResult, error) {
+	var hashes [][]byte
+	var hashesInitialized bool
+
+	// get a list of conditions (like "tx.height > 5")
+	conditions := q.Conditions()
+
+	// if there is a hash condition, return the result immediately
+	hash, err, ok := lookForHash(conditions)
+	if err != nil {
+		return nil, errors.Wrap(err, "error during searching for a hash in the query")
+	} else if ok {
+		res, err := txi.Get(hash)
+		return []*types.TxResult{res}, errors.Wrap(err, "error while retrieving the result")
+	}
+
+	// conditions to skip because they're handled before "everything else"
+	skipIndexes := make([]int, 0)
+
+	// if there is a height condition ("tx.height=3"), extract it for faster lookups
+	height, heightIndex := lookForHeight(conditions)
+	if heightIndex >= 0 {
+		skipIndexes = append(skipIndexes, heightIndex)
+	}
+
+	// extract ranges
+	// if both upper and lower bounds exist, it's better to get them in order not
+	// no iterate over kvs that are not within range.
+	ranges, rangeIndexes := lookForRanges(conditions)
+	if len(ranges) > 0 {
+		skipIndexes = append(skipIndexes, rangeIndexes...)
+
+		for _, r := range ranges {
+			if !hashesInitialized {
```

Comment:
this messes the whole code

Reply:
This looks much nicer. Can you clarify your comment? Does this still `mess the whole code` ? :P

Reply:
it is still not as elegant as it could? be, but I guess it's fine